### ISSUE TITLE
Extend RiboViz to use UMI-tools for demultiplexing

### ISSUE THREAD

**Issue by mikej888** [riboviz/RiboViz#84]
Follow on from #58

Demultiplexing needs barcode names and samples, which can be specified in the YAML file.

**Comment by mikej888** [comment]
Non-multiplexed data:

* Loop over samples for cutadapt + extract as at present.

Multiplexed data:

* `cutadapt`
* `umi_tools extract` barcodes into fastq headers.
* Given sample name and barcode, generate sample-specific fastq file names.
* Split fastq file into N fastq files.
* Loop over samples as at preset but skip cutadapt + extract.

Options for splitting fastq files:

* Edward's code:
  - Make sure it takes barcode out of header not out of read.
  - Ingest pyRNATagSeq demultiplexer and deduplicator code #26
* barcode-splitter
  - https://pypi.org/project/barcode-splitter/
  - https://anaconda.org/bioconda/barcode_splitter
  - conda install -c bioconda barcode_splitter
  - https://www.biostars.org/p/382142/
  - Recommended by Premal.
  - Ensure it does mismatches.
* Compare correctness and performance.

**Comment by kavousan** [comment]
@mikej888 cleaned up the code and renamed it. He now has a version working on own dummy data and it seems to group them correctly. To add a data-point that ought to go into unassigned (invalid barcode).

**Comment by mikej888** [comment]
When integrated and running on dummy data, functionality should be tested on yeast data from #39.

**Comment by mikej888** [comment]
Commits 921f4cb to 8419aa8 are ingest of SplitRNATagSeqFastQ.py from https://github.com/ewallace/pyRNATagSeq, with updates to be Python 3-compliant,  renamed to demultiplex_fastq.py, modularised into functions, updated with suggestions from pycodestyle and pylint, and modified to extract barcodes from fastq headers.

See 

* [riboviz/tools/demultiplex_fastq.py](https://github.com/riboviz/RiboViz/blob/use-umi-tools-deplex-84/riboviz/tools/demultiplex_fastq.py) 
* [docs/demultiplex-fastq.md](https://github.com/riboviz/RiboViz/blob/use-umi-tools-deplex-84/docs/demultiplex-fastq.md)

**Comment by mikej888** [comment]
See #95 for future task about Preprocessing sample file in demultiplex_fastq.py to address known issue (documented in code and docs) about mismatches and Hamming distances.

**Comment by mikej888** [comment]
@ewallace, questions relating to prep_riboviz.py:

* Should the workflow be able to handle a configuration which specifies both multiplexed and non-multiplexed sample files?
* Should the workflow be able to handle a configuration which specifies multiple multiplexed sample files?
* If yes to one or both of the above, is collation of TPMs applied across all the results from all the sample files? (i.e. given N multiplexed files only one TPMs_collated.tsv is produced) Or is collation applied to the results for each sample file in turn (i.e. given N multiplexed files N TPMS_collated.tsv files are produced)?

**Comment by ewallace** [comment]
>Should the workflow be able to handle a configuration which specifies both multiplexed and non-multiplexed sample files?

No.

>Should the workflow be able to handle a configuration which specifies multiple multiplexed sample files?

Yes please, but using the same barcodes for both, which is basically issue #28. We don't need to keep the outputs separate, i.e. it's ok to take 1/2/3/4 input fastq files and always output one fastq per barcode at the earliest possible stage after demultiplexing. So in fact we could `cat` the input fastqs as early as possible.

More complicated collating of results from multiple dissimilar samples is not a feature we need to add. It's ok for users to run RiboViz multiple times, then merge data from multiple samples themselves post-run using `collate_TPMs.R` or analogous tidyverse approaches.

**Comment by mikej888** [comment]
@ewallace, thanks. To summarise into a possible design:

For non-multiplexed data:

* Configuration:

```yaml
fq_files:
  SampleX: .fastq.gz, .fastq.gz
  SampleY: .fastq.gz, .fastq.gz
  ...
```

* Processing:

```
FOREACH sample
  FOREACH .fastq.gz
    cutadapt
    umi_tools extract
  Concatenate output .fastq.gz files into sample_complete.fastq.gz
  Carry out remaining sample-specific processing on sample_complete.fastq.gz
collate_tpms
```

Multiplexed data:

* Configuration:

```yaml
barcodes: .tsv
multiplexed_files:
- .fastq.gz
  .fastq.gz
  ...
```

* Processing:

```
FOREACH .fastq.gz
  cutadapt
  umi_tools extract
  demultiplex_fastq
FOREACH barcode:
  Concatenate demultiplexed .fastq.gz files into barcode.fastq.gz
  Carry out remaining sample-specific processing on barcode.fastq.gz
  collate_tpms
```

**Comment by ewallace** [comment]
What I would do is concatenate the .fastq.gz files into the first next processing step, for either by-sample or multiplexed processing. That produces bigger intermediate files, but reduces the explosion of intermediate files that, I think, do nothing useful.

```
cat part1.fastq.gz part2.fastq.gz part3.fastq.gz | \
cutadapt
```

I am not sure how to express this in the FOREACH pseudocode style above. Basically the goal is to *avoid* looping over the multiple input fastq.gz's, instead treating them as a single object as soon as possible.

**Comment by mikej888** [comment]
Cool, I see what is required.

**Comment by mikej888** [comment]
Up to commit 3f1f453 includes refactoring of `demultiplex_fastq.py`, additional helper functions to `utils.py` and, most significantly, changes to `create_fastq_examples.py` to create example data files for testing demultiplexing. See [create-fastq-examples.md](https://github.com/riboviz/RiboViz/blob/use-umi-tools-deplex-84/docs/create-fastq-examples.md)

**Comment by mikej888** [comment]
Example run of `cutadapt`, `umi_tools extract` and `demultiplex_fastq` using the example data from `create_fastq_examples.py` and checking the results against the example results:

```console
$ cutadapt --trim-n -O 1 -m 5 -a CTGTAGGCACC -o cut.fastq \
    data/example/example_multiplex_umi_barcode_adaptor.fastq

...
=== Summary ===

Total reads processed:                      90
Reads with adapters:                        90 (100.0%)
Reads that were too short:                   0 (0.0%)
Reads written (passing filters):            90 (100.0%)
...

$ cmp data/example/example_multiplex_umi_barcode.fastq cut.fastq 
$ echo $?
0

$ umi_tools extract --bc-pattern="^(?P<umi_1>.{4}).+(?P<umi_2>.{4})(?P<cell_1>.{3})$" \
    --extract-method=regex \
    -I cut.fastq \
    -S extract.fastq

...
2019-11-13 07:34:03,770 INFO Starting barcode extraction
2019-11-13 07:34:03,776 INFO Input Reads: 90
2019-11-13 07:34:03,776 INFO regex matches read1: 90
2019-11-13 07:34:03,776 INFO Reads output: 90
...

$ cmp data/example/example_multiplex.fastq extract.fastq
$ echo $?
0
```

Demultiplex `.fastq` file:

```console
$ python -m riboviz.tools.demultiplex_fastq \
    -r1 extract.fastq \
    -ss data/example/example_multiplex_barcodes.tsv \
    -o deplex/ \
    -m 2
$ ls -1 deplex/
example_multiplex_tag0.fastq
example_multiplex_tag1.fastq
example_multiplex_tag2.fastq
num_reads.tsv
Unassigned.fastq
$ cat deplex/num_reads.tsv 
SampleID	TagRead	NumReads
example_multiplex_tag0	ACG	27
example_multiplex_tag1	GAC	27
example_multiplex_tag2	CGA	27
Unassigned	NNNNNNNNN	9
Total		90
$ cat deplex/example_multiplex_tag0.fastq | grep Sim
@EWSim-1.1-umi5-reada-umix-bar0.0_ACG_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar0.0_ACG_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar0.0_ACG_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar0.0_ACG_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar0.0_ACG_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar0.0_ACG_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar0.0_ACG_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar0.0_ACG_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar0.0_ACG_CCCCCGTA
@EWSim-1.1-umi5-reada-umix-bar0.1_ACT_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar0.1_ACT_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar0.1_ACT_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar0.1_ACT_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar0.1_ACT_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar0.1_ACT_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar0.1_ACT_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar0.1_ACT_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar0.1_ACT_CCCCCGTA
@EWSim-1.1-umi5-reada-umix-bar0.2_TAG_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar0.2_TAG_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar0.2_TAG_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar0.2_TAG_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar0.2_TAG_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar0.2_TAG_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar0.2_TAG_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar0.2_TAG_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar0.2_TAG_CCCCCGTA
```

All barcodes belong to `0.x` group (for barcode `ACG`) as expected.

```console
$ cat deplex/example_multiplex_tag1.fastq | grep Sim
@EWSim-1.1-umi5-reada-umix-bar1.0_GAC_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar1.0_GAC_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar1.0_GAC_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar1.0_GAC_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar1.0_GAC_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar1.0_GAC_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar1.0_GAC_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar1.0_GAC_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar1.0_GAC_CCCCCGTA
@EWSim-1.1-umi5-reada-umix-bar1.1_GTC_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar1.1_GTC_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar1.1_GTC_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar1.1_GTC_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar1.1_GTC_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar1.1_GTC_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar1.1_GTC_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar1.1_GTC_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar1.1_GTC_CCCCCGTA
@EWSim-1.1-umi5-reada-umix-bar1.2_GTA_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar1.2_GTA_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar1.2_GTA_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar1.2_GTA_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar1.2_GTA_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar1.2_GTA_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar1.2_GTA_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar1.2_GTA_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar1.2_GTA_CCCCCGTA
```

All barcodes belong to `1.x` group (for barcode `GAC`) as expected.

```console
$ cat deplex/example_multiplex_tag2.fastq | grep Sim
@EWSim-1.1-umi5-reada-umix-bar2.0_CGA_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar2.0_CGA_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar2.0_CGA_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar2.0_CGA_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar2.0_CGA_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar2.0_CGA_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar2.0_CGA_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar2.0_CGA_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar2.0_CGA_CCCCCGTA
@EWSim-1.1-umi5-reada-umix-bar2.1_TGA_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar2.1_TGA_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar2.1_TGA_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar2.1_TGA_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar2.1_TGA_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar2.1_TGA_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar2.1_TGA_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar2.1_TGA_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar2.1_TGA_CCCCCGTA
@EWSim-1.1-umi5-reada-umix-bar2.2_CTT_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar2.2_CTT_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar2.2_CTT_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar2.2_CTT_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar2.2_CTT_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar2.2_CTT_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar2.2_CTT_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar2.2_CTT_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar2.2_CTT_CCCCCGTA
```

All barcodes belong to `2.x` group (for barcode `CGA`) as expected.

```console
$ cat deplex/Unassigned.fastq | grep Sim
@EWSim-1.1-umi5-reada-umix-bar3.0_TTT_AAAACGTA
@EWSim-1.2-umi5-reada-umix-bar3.0_TTT_AAAACGTA
@EWSim-1.3-umi5-readae-umix-bar3.0_TTT_AAAACGTA
@EWSim-2.1-umi5-reada-umiy-bar3.0_TTT_AAAAATAT
@EWSim-3.1-umi5-readb-umix-bar3.0_TTT_AAAACGTA
@EWSim-4.1-umi5-readb-umiz-bar3.0_TTT_AAAACGGC
@EWSim-4.2-umi5-readb-umiz-bar3.0_TTT_AAAACGGC
@EWSim-4.3-umi5-readb-umize-bar3.0_TTT_AAAACTGC
@EWSim-5.1-umi5c-readb-umix-bar3.0_TTT_CCCCCGTA
```

All barcodes belong to `3.x` group (for barcode `TTT`) as expected.

```console
$ cmp data/example/example_multiplex_tag0.fastq deplex/example_multiplex_tag0.fastq
$ echo $?
0
$ cmp data/example/example_multiplex_tag1.fastq deplex/example_multiplex_tag1.fastq
$ echo $?
0
$ cmp data/example/example_multiplex_tag2.fastq deplex/example_multiplex_tag2.fastq
$ echo $?
0
$ cmp data/example/example_multiplex_unassigned.fastq deplex/Unassigned.fastq
$ echo $?
0
$ cmp data/example/example_multiplex_num_reads.tsv deplex/num_reads.tsv 
$ echo $?
0
```

All results as expected.

Demultiplex `.fastq.gz` file:

```console
$ gzip -k extract.fastq
$ python -m riboviz.tools.demultiplex_fastq \
    -r1 extract.fastq.gz \
    -ss data/example/example_multiplex_barcodes.tsv \
    -o deplex-gz/ \
    -m 2
$ ls -1 deplex-gz
example_multiplex_tag0.fastq.gz
example_multiplex_tag1.fastq.gz
example_multiplex_tag2.fastq.gz
num_reads.tsv
Unassigned.fastq.gz
$ cat deplex-gz/num_reads.tsv 
SampleID	TagRead	NumReads
example_multiplex_tag0	ACG	27
example_multiplex_tag1	GAC	27
example_multiplex_tag2	CGA	27
Unassigned	NNNNNNNNN	9
Total		90
```

Compare to results from demultiplexing `.fastq file:

```console
$ gunzip deplex-gz/*gz
$ diff -rq deplex-gz deplex
$ echo $?
0
```

All results as expected.

**Comment by mikej888** [comment]
Up to commit 47cefbc adds work towards supporting demultiplexing a single file in `prep_riboviz.py`.

Added `vignette_multiplex_example_config.yaml` which uses includes `multiplex_fq_files`  and `barcodes` parameters, pointing to `data/examples/multiplex_umi_barcode_adaptor.fastq` and `data/examples/multiplex_barcodes.tsv` respectively.

Significant refactoring of `prep_riboviz.py` to extract out code that prepares each command submitted to bash into their own functions.

Added demultiplexing-specific code:

* Looks for `barcodes` and `multiplex_fq_files` configuration parameters and exits if either both these and `fq_files` are present or if neither these nor `fq_files` can be found.
* Processes samples as previously if `fq_files` is provided.
* Demultiplexes file if `multiplex_fq_files` and `barcodes` parameters are provided.
* Checks that files specified in `multiplex_fq_files`  and `barcodes` parameters exist.
* Invokes `cutadapt`, `umi_tools extract`, `demultiplex_fastq`.
* Gets list of samples which had one or more reads produced, from `num_reads.txt` output by `demultiplex_fast`, derives filenames and invokes `process_samples`.

Run of work to date:

```console
$ python -m riboviz.tools.prep_riboviz rscripts/ vignette/vignette_example_multiplex_config.yaml
...

WIP: multiplexed file processing
Processing file: data/example/multiplex_umi_barcode_adaptor.fastq
Cut Illumina adapters. Log: vignette/logs/20191115-100441/cutadapt.log
Extract barcodes and UMIs. Log: vignette/logs/20191115-100441/umi_tools_extract.log
Demultiplex reads. Log: vignette/logs/20191115-100441/demultiplex_fastq.log
['tag0', 'tag1', 'tag2']
{'tag0': 'tag0.fastq', 'tag1': 'tag1.fastq', 'tag2': 'tag2.fastq'}
Processing samples
Processing sample: tag0
Processing file: vignette/tmp/multiplex_umi_barcode_adaptor_deplex/tag0.fastq
Skipping adaptor trimming and barcode/UMI extraction
...
Create summary statistics and analyses plots. Log: vignette/logs/20191115-100722/tag0_09_generate_stats_figs.log
Finished processing sample: vignette/tmp/multiplex_umi_barcode_adaptor_deplex/tag0.fastq
Processing sample: tag1
Processing file: vignette/tmp/multiplex_umi_barcode_adaptor_deplex/tag1.fastq
Skipping adaptor trimming and barcode/UMI extraction
...
Create summary statistics and analyses plots. Log: vignette/logs/20191115-100722/tag1_09_generate_stats_figs.log
Finished processing sample: vignette/tmp/multiplex_umi_barcode_adaptor_deplex/tag1.fastq
Processing sample: tag2
Processing file: vignette/tmp/multiplex_umi_barcode_adaptor_deplex/tag2.fastq
Skipping adaptor trimming and barcode/UMI extraction
...
Create summary statistics and analyses plots. Log: vignette/logs/20191115-100722/tag2_09_generate_stats_figs.log
Finished processing sample: vignette/tmp/multiplex_umi_barcode_adaptor_deplex/tag2.fastq
Finished processing 3 samples, 0 failed
Completed
```

Check `cutadapt`, `umi_tools extract`, `demultiplex_fastq` outputs against expected outputs:

```console
$ cmp data/example/multiplex_umi_barcode.fastq vignette/tmp/multiplex_umi_barcode_adaptor_trim.fq 
$ echo $?
0
$ cmp data/example/multiplex.fastq vignette/tmp/multiplex_umi_barcode_adaptor_extract_trim.fq 
$ echo $?
0
$ diff -rq data/example/deplex/ vignette/tmp/multiplex_umi_barcode_adaptor_deplex/
$ echo $?
0
```

OK

Compare TPMs for reads against those output from a run of `vignette_example_config.yaml` (as the multiplexed data is based on the same reads):

```console
$ cmp vignette-dedup/output/umi5_umi3_tpms.tsv vignette/output/tag0_tpms.tsv
$ echo $?
0
$ cmp vignette-dedup/output/umi5_umi3_tpms.tsv vignette/output/tag1_tpms.tsv
$ echo $?
0
$ cmp vignette-dedup/output/umi5_umi3_tpms.tsv vignette/output/tag2_tpms.tsv
$ echo $?
0
```

OK

**Comment by mikej888** [comment]
@ewallace:

What about collate_tpms? Should this be done for each individual sample file produced by demultiplexing?

The "dry run" feature does not currently work if using demultiplexing. This is because the sample-specific fastq files which are to be processed are derived by parsing the `num_reads.tsv` file output from `demultiplex_fastq.py` and identifying the samples which have 1+ reads.

An option is, if doing a dry run, to use the barcodes file (e.g. `multiplex_barcodes.tsv`) that `demultiplex_fastq.py` takes as an input to determine the names of the sample files. This would mean a dry run bash script would not necessarily match a bash script produced by a "live" run if, during the "live" run, some sample-specific fastq files had 0 reads. Also, whether such a a dry run bash script would successfully run standalone also depends on how downstream processing steps react to the presence of empty sample-specific fastq files.

**Comment by mikej888** [comment]
TODOs for the record:

* Clean up the highly-proof-of-concept (!) code in `prep_riboviz.py`.
* Extend `test_prep_riboviz_example.py` introduce new `test_prep_riboviz_example_multiplex.py`.
* Extend `test_prep_riboviz.py` to test error conditions.
* Implement dry run fix.
* Refactor numbering of log files 1,2,2.1,2.2,2.3 etc.
* Update `prep_riboviz.py` header comment and user doc.

**Comment by ewallace** [comment]
@mikej888 

> What about collate_tpms? Should this be done for each individual sample file produced by demultiplexing?

Yes, exactly

> An option is, if doing a dry run, to use the barcodes file (e.g. multiplex_barcodes.tsv) that demultiplex_fastq.py takes as an input to determine the names of the sample files. This would mean a dry run bash script would not necessarily match a bash script produced by a "live" run if, during the "live" run, some sample-specific fastq files had 0 reads. Also, whether such a a dry run bash script would successfully run standalone also depends on how downstream processing steps react to the presence of empty sample-specific fastq files.

Yes, I think this is the right approach. In well-specified well-behaved data, the empty demultiplexed fastq file should not happen, but it might.

You have identified a possible bug. Right now we test and handle missing fastq files, but I don't know what happens if we have an input or demultiplexed fastq file that has 0 reads. I've created a new issue #97 to catch these and similar bugs.

**Comment by mikej888** [comment]
Up to commit 0628211 includes:

* Cleaned up code in `prep_riboviz.py`.
* Implemented fix so that, if `--dry-run` is enabled, then sample IDs, as specified in sample sheet, are used to identify names of demultiplexed fastq files whereas during a live run the `num_reads.tsv` file produced by `demultiplex_fastq.py` is used to identify names of fastq files with +1 reads.
* Added `test_prep_riboviz_example_multiplex.py` to run `example_multiplex_config.yaml` and validate post-trimmed fastq file, post-barcode and UMI-extracted fastq file, `num_reads.tsv` and fastq files produced by demultiplexing, and, for each fq file, post-deduplication UMI groups and finale `<SAMPLE>_TPMs_collated.tsv` files.
* Updated `prep_riboviz.py` comments, updated [docs/run-vignette.md](https://github.com/riboviz/RiboViz/blob/use-umi-tools-deplex-84/docs/run-vignette.md) on what `prep_riboviz.py` does and added a new page [docs/dedup-demultiplex.md](https://github.com/riboviz/RiboViz/blob/use-umi-tools-deplex-84/docs/dedup-demultiplex.md) describing how `prep_riboviz.py` can be configured for deduplication and demultiplexing and how it behaves.

**Comment by ewallace** [comment]
The documentation [docs/dedup-demultiplex.md](https://github.com/riboviz/RiboViz/blob/use-umi-tools-deplex-84/docs/dedup-demultiplex.md) is great. It might be better to have relative paths instead of, `vignette`,  `vignette/tmp`, etc., because on real data that directory would be called something else. Current wording makes it unclear how general the situation is.

**Comment by mikej888** [comment]
Up to commit 44bf959 includes:

* Updates to documentation to state paths are relative to dir_in|tmp|out|logs` .
* Added "--samples_in_sub_dirs=TRUE|FALSE" flag to collate_tpms.R.
 * Updated riboviz/workflow.py, collate_tpms to support are_samples_in_sub_dirs parameter.
* Updated prep_riboviz.py to call above function with value False or True depending on whether or not sample results are in sample-specific subdirectories used for demultiplexed samples (sample-specific subdirectories have not yet been implemented for non-demultiplexed samples - see #100)
* Updated prep_riboviz.py to collate TPMs across all demultiplexed samples, not once per sample.

**Comment by mikej888** [comment]
#100 makes handling of samples provided via `fq_files` parameter and samples arising from demultiplexing of file specified in `multiplex_fq_files` parameter consistent - sample-specific subdirectories of `tmp`, `logs`, and `output` are used.

@FlicAnderson to set up a copy of the regression test directory with sample-specific sub-directories:

```console
cp -r vignette-20191121-864be15/ vignette-20191121-864be15-sample-subdirs
cd vignette-20191121-864be15-sample-subdirs
mkdir tmp/WT3AT tmp/WTnone
mv tmp/WT3AT* tmp/WT3AT/
mv tmp/WTnone* tmp/WTnone

mkdir output/WT3AT output/WTnone
mv output/WT3AT* output/WT3AT/
mv output/WTnone* output/WTnone
```

**Comment by ewallace** [comment]
I just pushed to this branch in commit `f480dd8`, simdata renamed from example.

Simulated fastq data for umis/barcodes/adaptors is now called "simdata" and not "example".
That is because it is not a real example, but actually a simulation.
We may want to create "example" config files indicating how to deal with new data, rather than tied to the simulated data.
Also, build_indices=FALSE for simdata, because rebuilding indices for a tiny test is a waste of time.

I hope I didn't break anything!

**Comment by mikej888** [comment]
@ewallace, apologies I should have discussed the renaming before making the change. a864dad completes the required changes.

I've restored `build_indices=TRUE` for simdata_umi_config.yaml and simdata_multiplex_config.yaml as there is no guarantee that the indices will exist when the simulated data configurations are first run (which is the case when running tests using a fresh checkout). If users don't want to rebuild them they can set the flag to `FALSE` locally.

**Comment by mikej888** [comment]
Closed by #117
